Cubical/Foundations/Cubes/Base.agda

```diff
@@ -0,0 +1,155 @@
+{-
+
+This file contains:
+
+- The definition of the type of n-cubes;
+
+- Some basic operations.
+
+-}
+{-# OPTIONS --safe #-}
+module Cubical.Foundations.Cubes.Base where
+
+open import Cubical.Foundations.Prelude  hiding (Cube)
+open import Cubical.Foundations.Function hiding (const)
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Isomorphism
+
+open import Cubical.Data.Nat.Base
+open import Cubical.Data.Sigma.Base
+
+private
+  variable
+    ℓ : Level
+    A : Type ℓ
+
+
+-- The Type of n-Cubes
+
+-- P.S.
+-- Only the definitions of `∂Cube` and `CubeRel` essentially use mutual recursion.
+-- The case of `Cube` is designed to gain more definitional equality.
+
+interleaved mutual
+
+  Cube    : (n : ℕ) (A : Type ℓ) → Type ℓ
+  ∂Cube   : (n : ℕ) (A : Type ℓ) → Type ℓ
+  CubeRel : (n : ℕ) (A : Type ℓ) → ∂Cube n A → Type ℓ
+
+  Cube    0 A = A
+  Cube    (suc n) A = Σ[ ∂ ∈ ∂Cube (suc n) A ] CubeRel (suc n) A ∂
+
+  ∂Cube   0 A = Unit*
+  ∂Cube   1 A = A × A
+  ∂Cube   (suc (suc n)) A = Σ[ a₀ ∈ Cube (suc n) A ] Σ[ a₁ ∈ Cube (suc n) A ] a₀ .fst ≡ a₁ .fst
+
+  CubeRel 0 A _ = A
+  CubeRel 1 A ∂ = ∂ .fst ≡ ∂ .snd
+  CubeRel (suc (suc n)) A (a₀ , a₁ , ∂₋) = PathP (λ i → CubeRel (suc n) A (∂₋ i)) (a₀ .snd) (a₁ .snd)
+
+
+-- Some basic operations
+
+∂_ : {n : ℕ}{A : Type ℓ} → Cube n A → ∂Cube n A
+∂_ {n = 0} _ = tt*
+∂_ {n = suc n} = fst
+
+∂₀ : {n : ℕ}{A : Type ℓ} → Cube (suc n) A → Cube n A
+∂₀ {n = 0} (_ , p) = p i0
+∂₀ {n = suc n} (_ , p) = _ , p i0
+
+∂₁ : {n : ℕ}{A : Type ℓ} → Cube (suc n) A → Cube n A
+∂₁ {n = 0} (_ , p) = p i1
+∂₁ {n = suc n} (_ , p) = _ , p i1
+
+∂ᵇ₀ : {n : ℕ}{A : Type ℓ} → ∂Cube (suc n) A → Cube n A
+∂ᵇ₀ {n = 0} (a₀ , a₁) = a₀
+∂ᵇ₀ {n = suc n} (a₀ , a₁ , ∂₋) = a₀
+
+∂ᵇ₁ : {n : ℕ}{A : Type ℓ} → ∂Cube (suc n) A → Cube n A
+∂ᵇ₁ {n = 0} (a₀ , a₁) = a₁
+∂ᵇ₁ {n = suc n} (a₀ , a₁ , ∂₋) = a₁
+
+
+make∂ : {n : ℕ}{A : Type ℓ}{∂₀ ∂₁ : ∂Cube n A} → ∂₀ ≡ ∂₁ → CubeRel n A ∂₀ → CubeRel n A ∂₁ → ∂Cube (suc n) A
+make∂ {n = 0} _ a b = a , b
+make∂ {n = suc n} ∂₋ a₀ a₁ = (_ , a₀) , (_ , a₁) , ∂₋
+
+makeCube : {n : ℕ}{A : Type ℓ}{a₀ a₁ : Cube n A} → a₀ ≡ a₁ → Cube (suc n) A
+makeCube {n = 0} a₋ = _ , a₋
+makeCube {n = suc n} a₋ = _ , λ i → a₋ i .snd
+
+-- A cube is just a path of cubes of one-lower-dimension.
+-- Unfortunately the following function cannot begin at 0,
+-- because Agda doesn't support pattern matching on ℕ towards pre-types.
+pathCube : (n : ℕ) → (I → Cube (suc n) A) → Cube (suc (suc n)) A
+pathCube n p = _ , λ i → p i .snd
```

Comment:
Let's add a reminder here that we can fix this in 2.6.3 when I -> A becomes fibrant.

Reply:
done